src/zennit/rules.py

```diff
@@ -409,3 +409,30 @@ class ReLUGuidedBackprop(Hook):
     def backward(self, module, grad_input, grad_output):
         '''Modify ReLU gradient according to GuidedBackprop :cite:p:`springenberg2015striving`.'''
         return (grad_input[0] * (grad_output[0] > 0.),)
+
+
+class ReLUBetaSmooth(Hook):
+    '''Modify ReLU gradient to smooth softplus gradient :cite:p:`dombrowski2019explanations`. Used to obtain meaningful
+    surrogate gradients to compute higher order gradients with ReLUs. Equivalent to changing the gradient to be the
+    (scaled) logistic function (sigmoid).
+
+    Parameters
+    ----------
+    beta_smooth: float, optional
+        The beta parameter for the softplus gradient (i.e. ``sigmoid(beta * input)``). Defaults to ``10``.
+    '''
+    def __init__(self, beta_smooth=10.):
+        super().__init__()
+        self.beta_smooth = beta_smooth
+
+    def copy(self):
+        '''Return a copy of this hook with the same beta parameter.'''
+        return self.__class__(beta_smooth=self.beta_smooth)
+
+    def forward(self, module, input, output):
+        '''Remember the input for the backward pass.'''
+        self.stored_tensors['input'] = input
+
+    def backward(self, module, grad_input, grad_output):
+        '''Modify ReLU gradient to the smooth softplus gradient :cite:p:`dombrowski2019explanations`.'''
+        return (torch.sigmoid(self.beta_smooth * self.stored_tensors['input'][0]) * grad_output[0],)
```

Comment:
The `backward` function only modifies the gradient of the ReLU to become the gradient of Softplus, which is Sigmoid.

![image](https://user-images.githubusercontent.com/15217558/191696697-ea144077-731f-4b47-a437-7a246de0b369.png)

I decided to implement this as a hook to only change the gradient in the backward pass to a *smooth estimate* of the real gradient, since this will keep the model output *exactly* as intended.

I initially added a Canonizer, but I felt it did not belong there due to the ultimately changed output.

Reply:
You are completely right. I ignored the fact that this is inside the backward function of the activation function itself. Thanks for the hint. 😄

---

> I initially added a Canonizer, but I felt it did not belong there due to the ultimately changed output.

I also do not know the perfect place for this. But I also though that this is less a "Rule" and more a canonizer. Since the description of canonizers is:

https://github.com/chr5tphr/zennit/blob/0c0ab8b8520911c59b54e1c692be1efd4f70f4b0/src/zennit/canonizers.py#L29

I think the question is: Is this more a rule or rather something that is necessary to apply other rules (a canonizer)?

Since it's neither, it's more the case of preference. Personally, I am fine with both.
In general, I think that the programmer himself is responsible for the fact that the model is twice differentiable, if he wants to calculate the second gradient.